### ISSUE TITLE
Assigning values after chained masking is not working as expected

### ISSUE THREAD

**Issue by flonick** [pytorch/pytorch#133456]
### 🐛 Describe the bug

When assigning values after two masking operations, the assigned values are ignored in PyTorch 2.4.0.

Minimal example:
```python
mask1 = torch.zeros(3)
mask1[True][True]=1.
print(mask1)
```

The results will be a tensor with 3 zeros.
When only applying one True mask the values get updated. 
I also make the example slightly more complicated but easier to understand what it’s useful.

```
mask1 = torch.zeros(3)
mask1[[True,True,False]]=1
print(mask1[[True,True,False]])
mask1[[True,True,False]][[True,False]]=2
print(mask1[[True,True,False]][[True,False]])
```

The first print will be as expected 
tensor([1., 1.])
but the second print unexpectedly 
tensor([1.])

### Versions

torch 2.4.0

**Comment by Lazarus42** [comment]
This is occurring because chained indexing like this changes a copy of the original tensor. This behavior follows [NumPy](https://numpy.org/doc/stable/user/basics.indexing.html#assigning-values-to-indexed-arrays) as stated in the PyTorch [Tensor View documentation](https://pytorch.org/docs/1.13/tensor_view.html).

This can be seen by running 
```
mask1 = torch.zeros(3)
mask1[[True, True, False]] = 1
print(mask1)
```
which shows that `mask1 == tensor([1., 1., 0.])`, i.e. the original tensor changed because this is an assignment through boolean (not chained) indexing.

However, writing out the chained indexing with an intermediate variable shows:
```
mask1 = torch.zeros(3)
temp = mask1[[True, True, False]]
temp[[True, False]] = 1
print(temp)
print(mask1)
print(temp.storage().data_ptr() == mask1.storage().data_ptr())
```
You will see that `temp == tensor([1., 0.]), mask1 == tensor([0., 0., 0.]), temp.storage().data_ptr() != mask1.storage().data_ptr()` because `mask1[[True, True, False]]` creates a copy of the tensor on which the assignment is performed.

**Comment by flonick** [comment]
Thanks for the insightful answer.
Would it make sense to change this behavior? I have the feeling that the current solution is logical to some point, but also inconsistent. ChatGPT is also a bit confused about this, which is a sign for not as straight forward behavior as it should be.

When calling mask1[[True, True, False]] creates a copy of a tensor, why does `mask1[[True, True, False]] = 1` change the original tensor? Probably because when it's assigned, the copy is not created yet.
So it could be implemented, that `mask1[[True,True,False]][[True,False]]=2` changes the original tensor.

**Comment by colesbury** [comment]
This is the intended behavior. Please use the forum to ask questions: https://discuss.pytorch.org/

Please see:
* https://pytorch.org/cppdocs/notes/tensor_indexing.html
* https://numpy.org/doc/stable/user/basics.indexing.html
